File: fsharp-2017/src/Day03.fs
module Day03

let intSqrt = float >> sqrt >> int
let minus1IfEven x = if x % 2 = 0 then x - 1 else x

/// Part 1 Observation:
/// min-dist is sqrt(x)/2 + 1 and max-dist (corners) is min-dist*2
let distTo1 n =
  let sqrt = (n - 1) |> intSqrt |> minus1IfEven
  let minDist = sqrt/2 + 1
  let eps = abs <| (n - sqrt*sqrt) % (minDist * 2) - minDist
  minDist + eps

/// Part 2 Observation:
/// Numbers to add can be determined via calculating index in inner square 
let nextSpiralNum (xs:ResizeArray<int>) =
  let cnt = xs.Count
  let sqrt = cnt |> intSqrt |> minus1IfEven
  let edgeLen = sqrt + 1
  let len' = cnt - sqrt*sqrt + 1
  let dToCnr' = len' % edgeLen
  let edges = (len' + edgeLen - 1)/edgeLen |> int 
  let isLast = dToCnr' = 0 && edges = 4

  // i0 is index 0 of inner square. i is index of inner val adjacent to val to compute
  let i0 = if sqrt = 1 then 0 else (sqrt - 2) * (sqrt - 2) - 1
  let eps = if isLast then edgeLen - 1 else if dToCnr' = 0 then 0 else dToCnr' - 1
  let i = i0 + (edgeLen - 2)*(if isLast || dToCnr' > 0 then edges - 1 else edges) + eps

  let v0is0 = (dToCnr' = 0 && not isLast) || dToCnr' = 1 || (dToCnr' = 2 && edges = 1)
  // v0 | v1 = val at i | v2 | v3 = val at 2 index back from val to compute
  let v0 = if v0is0 then 0 else if len' = 1 then 0 else xs.[i-1]
  let v1 = if len' = 1 then 0 else xs.[i]
  let v2 = if dToCnr' = 0 || (dToCnr' = edgeLen - 1 && edges < 4) then 0 else xs.[i+1]
  let v3 = if ((dToCnr' = 1 && edges > 1) || (dToCnr' = 2 && edges = 1)) then xs.[cnt - 2] else 0
  v0 + v1 + v2 + v3 + xs.[cnt - 1]

let spiralNumAfter n =
  let a = ResizeArray()
  a.Add(1)
  let mutable next = nextSpiralNum a
  while next < n do
    a.Add(next)
    next <- nextSpiralNum a
  next

let run (s:string) =
  let n = int s
  printfn "Day 3-1 answer: %A" <| distTo1 n
  printfn "Day 3-2 answer: %A" <| spiralNumAfter n

File: js-pipe-fkit_2017/src/day03.js
const sqrtFloor = x => ~~Math.sqrt(x)
const minus1IfEven = x => x%2 == 0? x - 1 : x

/// Part 1 Observation:
/// minDist is sr(x)/2 + 1 and maxDist (corners) is minDist*2
const distTo1 = n /*int*/ => {
  const sr = (n - 1) |> sqrtFloor |> minus1IfEven
  const minDist = ~~(sr/2) + 1
  const eps = Math.abs((n - sr*sr) % (minDist * 2) - minDist)
  return minDist + eps
}

/// Part 2 Observation:
/// Numbers to add can be determined via calculating index in inner square 
const nextSpiralNum = xs /*[int]*/ => {
  const cnt = xs.length
  const sr = cnt |> sqrtFloor |> minus1IfEven
  const edgeLen = sr + 1
  const len = cnt - sr*sr + 1
  const dToCnr = len % edgeLen
  const edges = (len + edgeLen - 1)/edgeLen |> Math.floor
  const isLast = dToCnr==0 && edges==4

  // i0 is index 0 of inner square. i is index of inner val adjacent to val to compute
  const i0 = sr==1? 0 : (sr - 2)**2 - 1
  const eps = isLast? (edgeLen - 1) : (dToCnr==0)? 0 : dToCnr - 1
  const i = i0 + (edgeLen - 2) * ((isLast || dToCnr > 0)? (edges - 1) : edges) + eps

  const v0is0 = (dToCnr==0 && !isLast) || (dToCnr==1) || (dToCnr==2 && edges==1)
  // v0 ; v1 = val at i ; v2 ; v3 = val at 2 index back from val to compute
  const v0 = v0is0? 0 : len==1? 0 : xs[i-1]
  const v1 = len==1? 0 : xs[i]
  const v2 = dToCnr==0 || ((dToCnr == edgeLen-1) && edges < 4)? 0 : xs[i+1]
  const v3 = ((dToCnr==1 && edges > 1) || (dToCnr==2 && edges==1))? xs[cnt - 2] : 0
  return v0 + v1 + v2 + v3 + xs[cnt - 1]
}

const spiralNumAfter = n /*int*/ => {
  const a = [1]
  let next = nextSpiralNum(a)
  while (next < n) {
    a.push(next)
    next = nextSpiralNum(a)
  }
  return next
}

const run = (inStr, log) => {
  const n = Number(inStr)
  log.p1(n |> distTo1)
  log.p2(n |> spiralNumAfter)
}
module.exports = { run }